File: cli/internal/tracker/tracker.go
// Package tracker provides a tracker integration abstraction and
// a Linear implementation for tyr-mini.
package tracker

// Project represents an external tracker project (Linear project).
type Project struct {
	ID             string  `json:"id"`
	Name           string  `json:"name"`
	Description    string  `json:"description"`
	Status         string  `json:"status"`
	URL            string  `json:"url"`
	MilestoneCount int     `json:"milestone_count"`
	IssueCount     int     `json:"issue_count"`
	Slug           string  `json:"slug"`
	Progress       float64 `json:"progress"`
	StartDate      *string `json:"start_date"`
	TargetDate     *string `json:"target_date"`
}

// Milestone represents a project milestone (Linear project milestone).
type Milestone struct {
	ID         string  `json:"id"`
	ProjectID  string  `json:"project_id"`
	Name       string  `json:"name"`
	Desc       string  `json:"description"`
	SortOrder  int     `json:"sort_order"`
	Progress   float64 `json:"progress"`
	TargetDate *string `json:"target_date"`
}

// Issue represents a tracker issue (Linear issue).
type Issue struct {
	ID            string   `json:"id"`
	Identifier    string   `json:"identifier"`
	Title         string   `json:"title"`
	Description   string   `json:"description"`
	Status        string   `json:"status"`
	StatusType    string   `json:"status_type"`
	Assignee      *string  `json:"assignee"`
	Labels        []string `json:"labels"`
	Priority      int      `json:"priority"`
	PriorityLabel string   `json:"priority_label"`
	Estimate      *int     `json:"estimate"`
	URL           string   `json:"url"`
	MilestoneID   *string  `json:"milestone_id"`
}

// Tracker is the port for external issue tracker integration.
// Implementations: LinearTracker.
// TODO(tracker): Add GitHub Issues, Jira adapters.
type Tracker interface {
	ListProjects() ([]Project, error)
	GetProject(id string) (*Project, error)
	GetProjectFull(id string) (*Project, []Milestone, []Issue, error)
	ListMilestones(projectID string) ([]Milestone, error)
	ListIssues(projectID string, milestoneID *string) ([]Issue, error)

	CreateProject(name, description string) (string, error)
	CreateMilestone(name, projectID string, sortOrder float64) (string, error)
	CreateIssue(title, description, projectID string, milestoneID *string, estimate *int) (string, error)

	UpdateIssueState(issueID, stateName string) error
	AddComment(issueID, body string) error

	Close() error
}

File: web-next/packages/plugin-tyr/src/domain/tracker.ts
import { z } from 'zod';

/**
 * Tracker browser domain types.
 *
 * The tracker browser lets operators browse projects, milestones, and issues
 * from an external issue tracker (Linear, GitHub Projects, etc.) and import
 * them as Sagas.
 *
 * Owner: plugin-tyr.
 */

export const trackerProjectSchema = z.object({
  id: z.string().min(1),
  name: z.string().min(1),
  description: z.string(),
  status: z.string(),
  url: z.string(),
  milestoneCount: z.number().int().nonnegative(),
  issueCount: z.number().int().nonnegative(),
});
export type TrackerProject = z.infer<typeof trackerProjectSchema>;

export const trackerMilestoneSchema = z.object({
  id: z.string().min(1),
  projectId: z.string().min(1),
  name: z.string().min(1),
  description: z.string(),
  sortOrder: z.number().int().nonnegative(),
  progress: z.number().min(0).max(100),
});
export type TrackerMilestone = z.infer<typeof trackerMilestoneSchema>;

export const trackerIssueSchema = z.object({
  id: z.string().min(1),
  identifier: z.string().min(1),
  title: z.string().min(1),
  description: z.string(),
  status: z.string(),
  assignee: z.string().nullable(),
  labels: z.array(z.string()),
  priority: z.number().int().nonnegative(),
  url: z.string(),
  milestoneId: z.string().nullable(),
});
export type TrackerIssue = z.infer<typeof trackerIssueSchema>;

export const repoInfoSchema = z.object({
  provider: z.string(),
  org: z.string(),
  name: z.string(),
  cloneUrl: z.string(),
  url: z.string(),
  defaultBranch: z.string(),
  branches: z.array(z.string()),
});
export type RepoInfo = z.infer<typeof repoInfoSchema>;
